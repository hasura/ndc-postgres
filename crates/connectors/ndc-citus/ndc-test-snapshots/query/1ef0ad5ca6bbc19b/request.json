{
  "collection": "citus_tables",
  "query": {
    "fields": {
      "access_method": {
        "type": "column",
        "column": "access_method"
      },
      "citus_table_type": {
        "type": "column",
        "column": "citus_table_type"
      },
      "colocation_id": {
        "type": "column",
        "column": "colocation_id"
      },
      "distribution_column": {
        "type": "column",
        "column": "distribution_column"
      },
      "shard_count": {
        "type": "column",
        "column": "shard_count"
      },
      "table_name": {
        "type": "column",
        "column": "table_name"
      },
      "table_owner": {
        "type": "column",
        "column": "table_owner"
      },
      "table_size": {
        "type": "column",
        "column": "table_size"
      }
    },
    "limit": 10,
    "order_by": {
      "elements": [
        {
          "order_direction": "desc",
          "target": {
            "type": "column",
            "name": "table_size",
            "path": []
          }
        },
        {
          "order_direction": "desc",
          "target": {
            "type": "column",
            "name": "colocation_id",
            "path": []
          }
        }
      ]
    }
  },
  "arguments": {},
  "collection_relationships": {}
}