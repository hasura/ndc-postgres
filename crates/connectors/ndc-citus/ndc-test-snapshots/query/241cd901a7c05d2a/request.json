{
  "collection": "Genre",
  "query": {
    "fields": {
      "GenreId": {
        "type": "column",
        "column": "GenreId"
      },
      "Name": {
        "type": "column",
        "column": "Name"
      }
    },
    "limit": 10,
    "where": {
      "type": "binary_comparison_operator",
      "column": {
        "type": "column",
        "name": "Name",
        "path": []
      },
      "operator": {
        "type": "equal"
      },
      "value": {
        "type": "scalar",
        "value": "Jazz"
      }
    }
  },
  "arguments": {},
  "collection_relationships": {}
}