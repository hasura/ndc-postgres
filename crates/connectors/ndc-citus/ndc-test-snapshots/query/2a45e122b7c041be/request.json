{
  "collection": "PlaylistTrack",
  "query": {
    "fields": {
      "PlaylistId": {
        "type": "column",
        "column": "PlaylistId"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      }
    },
    "limit": 10
  },
  "arguments": {},
  "collection_relationships": {}
}