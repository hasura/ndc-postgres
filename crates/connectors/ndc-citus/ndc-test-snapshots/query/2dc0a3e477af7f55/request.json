{
  "collection": "MediaType",
  "query": {
    "fields": {
      "MediaTypeId": {
        "type": "column",
        "column": "MediaTypeId"
      },
      "Name": {
        "type": "column",
        "column": "Name"
      }
    },
    "limit": 10,
    "order_by": {
      "elements": []
    }
  },
  "arguments": {},
  "collection_relationships": {}
}