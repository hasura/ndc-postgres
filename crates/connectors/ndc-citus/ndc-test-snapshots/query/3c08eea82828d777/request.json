{
  "collection": "PlaylistTrack",
  "query": {
    "fields": {
      "PlaylistId": {
        "type": "column",
        "column": "PlaylistId"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      }
    },
    "limit": 10,
    "where": {
      "type": "binary_comparison_operator",
      "column": {
        "type": "column",
        "name": "PlaylistId",
        "path": []
      },
      "operator": {
        "type": "equal"
      },
      "value": {
        "type": "scalar",
        "value": 1
      }
    }
  },
  "arguments": {},
  "collection_relationships": {}
}