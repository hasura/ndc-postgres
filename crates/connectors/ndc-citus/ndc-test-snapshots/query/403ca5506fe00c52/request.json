{
  "collection": "Track",
  "query": {
    "fields": {
      "AlbumId": {
        "type": "column",
        "column": "AlbumId"
      },
      "Bytes": {
        "type": "column",
        "column": "Bytes"
      },
      "Composer": {
        "type": "column",
        "column": "Composer"
      },
      "GenreId": {
        "type": "column",
        "column": "GenreId"
      },
      "MediaTypeId": {
        "type": "column",
        "column": "MediaTypeId"
      },
      "Milliseconds": {
        "type": "column",
        "column": "Milliseconds"
      },
      "Name": {
        "type": "column",
        "column": "Name"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      },
      "UnitPrice": {
        "type": "column",
        "column": "UnitPrice"
      }
    },
    "limit": 10,
    "where": {
      "type": "binary_comparison_operator",
      "column": {
        "type": "column",
        "name": "GenreId",
        "path": []
      },
      "operator": {
        "type": "equal"
      },
      "value": {
        "type": "scalar",
        "value": 1
      }
    }
  },
  "arguments": {},
  "collection_relationships": {}
}