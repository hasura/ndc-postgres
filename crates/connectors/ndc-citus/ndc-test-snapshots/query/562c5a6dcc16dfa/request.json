{
  "collection": "PlaylistTrack",
  "query": {
    "fields": {
      "PlaylistId": {
        "type": "column",
        "column": "PlaylistId"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      }
    },
    "limit": 10,
    "order_by": {
      "elements": [
        {
          "order_direction": "desc",
          "target": {
            "type": "column",
            "name": "PlaylistId",
            "path": []
          }
        }
      ]
    }
  },
  "arguments": {},
  "collection_relationships": {}
}