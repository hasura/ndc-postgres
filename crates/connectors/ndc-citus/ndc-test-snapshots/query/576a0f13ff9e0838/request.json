{
  "collection": "Track",
  "query": {
    "fields": {
      "AlbumId": {
        "type": "column",
        "column": "AlbumId"
      },
      "Bytes": {
        "type": "column",
        "column": "Bytes"
      },
      "Composer": {
        "type": "column",
        "column": "Composer"
      },
      "GenreId": {
        "type": "column",
        "column": "GenreId"
      },
      "MediaTypeId": {
        "type": "column",
        "column": "MediaTypeId"
      },
      "Milliseconds": {
        "type": "column",
        "column": "Milliseconds"
      },
      "Name": {
        "type": "column",
        "column": "Name"
      },
      "TrackId": {
        "type": "column",
        "column": "TrackId"
      },
      "UnitPrice": {
        "type": "column",
        "column": "UnitPrice"
      }
    },
    "limit": 10,
    "order_by": {
      "elements": [
        {
          "order_direction": "desc",
          "target": {
            "type": "column",
            "name": "Composer",
            "path": []
          }
        }
      ]
    }
  },
  "arguments": {},
  "collection_relationships": {}
}