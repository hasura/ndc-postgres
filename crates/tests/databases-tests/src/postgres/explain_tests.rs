#[cfg(test)]
mod query {
    use super::super::common::create_router;
    use tests_common::assert::is_contained_in_lines;
    use tests_common::request::run_query_explain;

    #[tokio::test]
    async fn select_by_pk() {
        let result = run_query_explain(create_router().await, "select_by_pk").await;
        is_contained_in_lines(&["Aggregate", "Scan", "35"], &result.details.plan);
        insta::assert_snapshot!(result.details.query);
    }

    #[tokio::test]
    async fn select_where_variable() {
        let result = run_query_explain(create_router().await, "select_where_variable").await;
        is_contained_in_lines(&["Aggregate", "Seq Scan", "Filter"], &result.details.plan);
        insta::assert_snapshot!(result.details.query);
    }

    #[tokio::test]
    async fn select_where_name_nilike() {
        let result = run_query_explain(create_router().await, "select_where_name_nilike").await;
        let keywords = &[
            "Aggregate",
            "Subquery Scan",
            "Limit",
            "Index Scan",
            "Filter",
        ];
        is_contained_in_lines(keywords, &result.details.plan);
        insta::assert_snapshot!(result.details.query);
    }

    #[tokio::test]
    async fn duplicate_filter_results() {
        let result = run_query_explain(create_router().await, "duplicate_filter_results").await;
        let keywords = &[
            "Aggregate",
            "Subquery Scan",
            "Limit",
            "Index Scan",
            "Filter",
        ];
        is_contained_in_lines(keywords, &result.details.plan);
        insta::assert_snapshot!(result.details.query);
    }

    #[tokio::test]
    async fn duplicate_filter_results_nested() {
        let result =
            run_query_explain(create_router().await, "duplicate_filter_results_nested").await;
        let keywords = &[
            "Aggregate",
            "Subquery Scan",
            "Limit",
            "Index Scan",
            "Filter",
        ];
        is_contained_in_lines(keywords, &result.details.plan);
        insta::assert_snapshot!(result.details.query);
    }

    #[tokio::test]
    async fn order_by_nested_field() {
        let result = run_query_explain(create_router().await, "order_by_nested_field").await;
        let keywords = &[
            "Sort Key",
            "ProjectSet",
            "group_leader\".characters).name) DESC",
        ];
        is_contained_in_lines(keywords, &result.details.plan);
        insta::assert_snapshot!(result.details.query);
    }

    mod native_queries {
        use super::super::super::common::create_router;
        use tests_common::assert::is_contained_in_lines;
        use tests_common::request::run_query_explain;

        #[tokio::test]
        async fn embedded_variable() {
            let result =
                run_query_explain(create_router().await, "native_queries/embedded_variable").await;
            is_contained_in_lines(&["Aggregate", "Seq Scan", "Filter"], &result.details.plan);
            insta::assert_snapshot!(result.details.query);
        }
    }

    #[tokio::test]
    async fn select_where_no_variable() {
        let result = run_query_explain(create_router().await, "select_where_no_variables").await;
        assert!(result.details.plan.is_empty());
        insta::assert_snapshot!(result.details.query);
    }

    #[tokio::test]
    async fn nested_field_relationship() {
        let result = run_query_explain(create_router().await, "nested_field_relationship").await;
        insta::assert_snapshot!(result.details.query);
    }
}

#[cfg(test)]
mod mutation {
    use super::super::common::create_router;
    use tests_common::assert::is_contained_in_lines;
    use tests_common::request::run_mutation_explain;

    #[tokio::test]
    async fn delete_playlist_track() {
        let result = run_mutation_explain(create_router().await, "delete_playlist_track").await;
        is_contained_in_lines(
            &["Delete", "Index Cond", "CTE Scan"],
            result
                .details
                .get("0 delete_playlist_track Execution Plan")
                .unwrap(),
        );
        insta::assert_snapshot!(result
            .details
            .get("0 delete_playlist_track SQL Mutation")
            .unwrap());
    }

    #[tokio::test]
    async fn insert_artist_album() {
        let result = run_mutation_explain(create_router().await, "insert_artist_album").await;
        is_contained_in_lines(
            &["Insert", "CTE Scan"],
            result
                .details
                .get("0 insert_artist Execution Plan")
                .unwrap(),
        );
        is_contained_in_lines(
            &["Insert", "CTE Scan"],
            result.details.get("1 insert_album Execution Plan").unwrap(),
        );
        let queries = [
            result.details.get("0 insert_artist SQL Mutation").unwrap(),
            "\n\n\n",
            result.details.get("1 insert_album SQL Mutation").unwrap(),
        ]
        .concat();
        insta::assert_snapshot!(queries);
    }

    #[tokio::test]
    async fn v2_insert_custom_dog() {
        let result = run_mutation_explain(create_router().await, "v2_insert_custom_dog").await;
        is_contained_in_lines(
            &["Insert", "Aggregate"],
            result
                .details
                .get("0 v2_insert_custom_dog Execution Plan")
                .unwrap(),
        );
        insta::assert_snapshot!(result
            .details
            .get("0 v2_insert_custom_dog SQL Mutation")
            .unwrap());
    }

    #[tokio::test]
    async fn delete_invoice_line() {
        let result = run_mutation_explain(create_router().await, "delete_invoice_line").await;
        is_contained_in_lines(
            &[
                "Delete",
                "Filter",
                "Index Cond",
                "Nested Loop",
                "Index Scan",
            ],
            result
                .details
                .get("0 v2_delete_InvoiceLine_by_InvoiceLineId Execution Plan")
                .unwrap(),
        );
        insta::assert_snapshot!(result
            .details
            .get("0 v2_delete_InvoiceLine_by_InvoiceLineId SQL Mutation")
            .unwrap());
    }

    #[tokio::test]
    async fn v2_insert_update_custom_dog() {
        let result =
            run_mutation_explain(create_router().await, "v2_insert_update_custom_dog").await;
        is_contained_in_lines(
            &["Update", "Aggregate"],
            result
                .details
                .get("1 v2_update_custom_dog_by_id Execution Plan")
                .unwrap(),
        );
        insta::assert_snapshot!(result
            .details
            .get("1 v2_update_custom_dog_by_id SQL Mutation")
            .unwrap());
    }
}
