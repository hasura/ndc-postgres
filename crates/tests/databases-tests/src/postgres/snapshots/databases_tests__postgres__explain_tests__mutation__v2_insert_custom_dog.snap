---
source: crates/tests/databases-tests/src/postgres/explain_tests.rs
expression: "result.details.get(\"0 insert_custom_dog SQL Mutation\").unwrap()"
---
EXPLAIN WITH "%0_generated_mutation" AS (
  INSERT INTO
    "custom"."dog"("adopter_name", "birthday", "height_cm", "name")
  VALUES
    (
      cast($1 as "pg_catalog"."text"),
      cast($2 as "pg_catalog"."date"),
      160,
      cast($3 as "pg_catalog"."text")
    ),
    (
      cast($4 as "pg_catalog"."text"),
      DEFAULT,
      160,
      cast($5 as "pg_catalog"."text")
    ) RETURNING *,
    (
      "custom"."dog"."adopter_name" = cast($6 as "pg_catalog"."text")
    ) AS "%check__constraint"
)
SELECT
  (
    SELECT
      json_build_object('result', row_to_json("%5_universe"), 'type', $7) AS "universe"
    FROM
      (
        SELECT
          *
        FROM
          (
            SELECT
              coalesce(json_agg(row_to_json("%6_returning")), '[]') AS "returning"
            FROM
              (
                SELECT
                  cast("%2_custom_dog"."id" as "text") AS "id",
                  "%2_custom_dog"."name" AS "name",
                  "%2_custom_dog"."birthday" AS "birthday",
                  cast("%2_custom_dog"."height_cm" as "text") AS "height_cm",
                  cast("%2_custom_dog"."height_in" as "text") AS "height_inch",
                  "%2_custom_dog"."adopter_name" AS "adopter_name"
                FROM
                  (
                    SELECT
                      "%1_custom_dog".*
                    FROM
                      "%0_generated_mutation" AS "%1_custom_dog"
                  ) AS "%2_custom_dog"
              ) AS "%6_returning"
          ) AS "%6_returning"
          CROSS JOIN (
            SELECT
              COUNT(*) AS "affected_rows"
            FROM
              (
                SELECT
                  "%3_custom_dog".*
                FROM
                  "%0_generated_mutation" AS "%3_custom_dog"
              ) AS "%4_custom_dog"
          ) AS "%7_aggregates"
      ) AS "%5_universe"
  ) AS "%results",
  (
    SELECT
      coalesce(
        bool_and("%8_insert_custom_dog"."%check__constraint"),
        true
      ) AS "%check__constraint"
    FROM
      "%0_generated_mutation" AS "%8_insert_custom_dog"
  ) AS "%check__constraint"
